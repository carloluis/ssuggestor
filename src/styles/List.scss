:global .ss-list {
	display: block;
	max-height: 250px;
	min-width: 30px;
	width: inherit;
	overflow: auto;
	cursor: pointer;
}
