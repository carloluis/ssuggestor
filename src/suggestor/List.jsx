import React from 'react';
import PropTypes from 'prop-types';
import ListItem from './ListItem';

const List = ({ open, filtered, index, value, onItemClick, onItemMouseEnter, classSchema }) =>
	open &&
	!!filtered.length && (
		<ul className={classSchema.list}>
			{filtered.map((item, i) => (
				<ListItem
					key={item.word}
					{...{
						item,
						onItemClick,
						onItemMouseEnter,
						search: value,
						index: i,
						overItem: i === index,
						classSchema
					}}
				/>
			))}
		</ul>
	);

List.propTypes = {
	filtered: PropTypes.array.isRequired,
	index: PropTypes.number.isRequired,
	onItemClick: PropTypes.func.isRequired,
	onItemMouseEnter: PropTypes.func.isRequired,
	value: PropTypes.string.isRequired,
	open: PropTypes.bool.isRequired,
	classSchema: PropTypes.shape({
		list: PropTypes.string,
		item: PropTypes.string,
		activeItem: PropTypes.string
	})
};

export default List;
