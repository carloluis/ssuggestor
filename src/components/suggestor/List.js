import React from 'react';
import ListItem from './ListItem';
import { getListStyles } from './styles';

export const List = ({ open, list, index, value, onItemClick, onItemMouseEnter }) => (
	(open && !!list.length) && <ul className="dropdown-menu" style={getListStyles(open)}>
		{ list.map((item, i) => <ListItem key={i} {...{ item, onItemClick, onItemMouseEnter, search:value, index:i, overItem:index===i }} />) }
	</ul>
);

List.propTypes = {
	list: React.PropTypes.arrayOf(React.PropTypes.string).isRequired,
	open: React.PropTypes.bool.isRequired,
	index: React.PropTypes.number.isRequired,
	onItemClick: React.PropTypes.func.isRequired,
	onItemMouseEnter: React.PropTypes.func.isRequired,
	value: React.PropTypes.string.isRequired
};

export default List;