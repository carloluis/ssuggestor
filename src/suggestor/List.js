import React from 'react';
import ListItem from './ListItem';
import { getListStyles } from './styles';

export const List = ({ open, filtered, index, value, onItemClick, onItemMouseEnter }) => (
	(open && !!filtered.length) && 
	<ul className="dropdown-menu" style={getListStyles(open)}>
		{ filtered.map((item, i) => 
			<ListItem key={item.word} {...{ item, onItemClick, onItemMouseEnter, search:value, index:i, overItem:i===index }} />) }
	</ul>
);

List.propTypes = {
	filtered: React.PropTypes.array.isRequired,
	index: React.PropTypes.number.isRequired,
	value: React.PropTypes.string.isRequired,
	open: React.PropTypes.bool.isRequired,
	onItemClick: React.PropTypes.func.isRequired,
	onItemMouseEnter: React.PropTypes.func.isRequired
};

export default List;