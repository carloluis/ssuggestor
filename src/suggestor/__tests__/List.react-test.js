import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import List from '../List';

describe('<List />', () => {
	let props;
	beforeEach(() => {
		props = {
			filtered: [],
			open: false,
			index: 0,
			onItemClick: _=>_,
			onItemMouseEnter: _=>_,
			value: ''
		};
	});
	it('render with no suggestions list', () => {
		const tree = renderer.create(<List {...props} />);
		expect(tree).toMatchSnapshot();
	});
	it('render with suggestions list (open:false)', () => {
		props = { ...props, filtered: [{ word:'suggest-1', index:0 }] };
		const tree = renderer.create(<List {...props} />);
		expect(tree).toMatchSnapshot();
	});
	it('render with suggestions list (open:true)', () => {
		props = { ...props, filtered: [{ word:'suggest-1', index:0 }], open:true };
		const tree = renderer.create(<List {...props} />);
		expect(tree).toMatchSnapshot();
	});
	it('render with suggestions list (open:true, index:0)', () => {
		props = { ...props, filtered: [{ word:'suggest-0', index:0 }, { word:'suggest-1', index:0 }], open:true };
		const tree = renderer.create(<List {...props} />);
		expect(tree).toMatchSnapshot();
	});
	it('render with suggestions list (open:true, index:1)', () => {
		props = { ...props, filtered: [{ word:'suggest-0', index:0 }, { word:'suggest-1', index:0 }], open:true, index:1 };
		const tree = renderer.create(<List {...props} />);
		expect(tree).toMatchSnapshot();
	});
});

test('List with open:false => no renders ul.dropdown-menu', () => {
	const props = { filtered: [], open: false, index: 0, onItemClick: f => f, onItemMouseEnter: f => f, value: '' };

	const wrapper = mount(<List {...props} />);

	expect(wrapper.find('.dropdown-menu').node).toBeFalsy();
});

test('List with empty list => no renders ul.dropdown-menu', () => {
	const props = { filtered: [], open: true, index: 0, onItemClick: f => f, onItemMouseEnter: f => f, value: '' };

	const wrapper = mount(<List {...props} />);

	expect(wrapper.find('.dropdown-menu').node).toBeFalsy();
});

test('List with open:true and not empty list => renders ul.dropdown-menu', () => {
	const props = { filtered: [{ word:'x', index:0 }], open: true, index: 0, onItemClick: f => f, onItemMouseEnter: f => f, value: '' };

	const wrapper = mount(<List {...props} />);

	expect(wrapper.find('.dropdown-menu').node).toBeTruthy();
});

describe('List Component', () => {
	let props;
	beforeEach(() => {
		props = {
			filtered: [],
			value: '',
			open: true,
			index: 0,
			onItemClick: jest.fn(),
			onItemMouseEnter: _=>_
		};
	});
	describe('List with item X', () => {
		const ITEM = { word: 'X', index: 0 };
		beforeEach(() => {
			props.filtered.push(ITEM);
		});

		it(`li's text should match item`, () => {
			const wrapper = mount(<List {...props} />);

			const li = wrapper.find('li');
			expect(li.text()).toBe(ITEM.word);
		});

		describe('when click on li', () => {
			it('call `onItemClick` callback', () => {
				const wrapper = mount(<List {...props} />);

				const li = wrapper.find('li');
				li.simulate('click');

				expect(props.onItemClick).toBeCalled();
			});

			it('call `onItemClick` callback with `item`', () => {
				const wrapper = mount(<List {...props} />);

				const li = wrapper.find('li');
				li.simulate('click');

				expect(props.onItemClick).toBeCalledWith(ITEM);
			});
		});
	});
})
