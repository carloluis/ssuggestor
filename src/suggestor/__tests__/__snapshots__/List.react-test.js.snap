exports[`<List /> render with no suggestions list 1`] = `null`;

exports[`<List /> render with suggestions list (open:false) 1`] = `null`;

exports[`<List /> render with suggestions list (open:true) 1`] = `
<ul
  className="dropdown-menu"
  style={
    Object {
      "cursor": "pointer",
      "display": "block",
      "maxHeight": 250,
      "minWidth": 30,
      "overflow": "auto",
      "width": "inherit",
    }
  }>
  <li
    onClick={[Function]}
    onMouseEnter={[Function]}
    style={
      Object {
        "backgroundColor": "#f5f5f5",
      }
    }
    value="suggest-1">
    <a>
      suggest-1
    </a>
  </li>
</ul>
`;

exports[`<List /> render with suggestions list (open:true, index:0) 1`] = `
<ul
  className="dropdown-menu"
  style={
    Object {
      "cursor": "pointer",
      "display": "block",
      "maxHeight": 250,
      "minWidth": 30,
      "overflow": "auto",
      "width": "inherit",
    }
  }>
  <li
    onClick={[Function]}
    onMouseEnter={[Function]}
    style={
      Object {
        "backgroundColor": "#f5f5f5",
      }
    }
    value="suggest-0">
    <a>
      suggest-0
    </a>
  </li>
  <li
    onClick={[Function]}
    onMouseEnter={[Function]}
    style={
      Object {
        "backgroundColor": false,
      }
    }
    value="suggest-1">
    <a>
      suggest-1
    </a>
  </li>
</ul>
`;

exports[`<List /> render with suggestions list (open:true, index:1) 1`] = `
<ul
  className="dropdown-menu"
  style={
    Object {
      "cursor": "pointer",
      "display": "block",
      "maxHeight": 250,
      "minWidth": 30,
      "overflow": "auto",
      "width": "inherit",
    }
  }>
  <li
    onClick={[Function]}
    onMouseEnter={[Function]}
    style={
      Object {
        "backgroundColor": false,
      }
    }
    value="suggest-0">
    <a>
      suggest-0
    </a>
  </li>
  <li
    onClick={[Function]}
    onMouseEnter={[Function]}
    style={
      Object {
        "backgroundColor": "#f5f5f5",
      }
    }
    value="suggest-1">
    <a>
      suggest-1
    </a>
  </li>
</ul>
`;
